### ISSUE TITLE
[Feature Request] Allow for multi-phase optimization in Trainer

### ISSUE THREAD

**Issue by bsprenger** [pytorch/rl#3469]
## Motivation

The `Trainer`'s optimization loop currently has a fixed structure:

```py
losses_td = self.loss_module(sub_batch)       # compute ALL losses up-front
self._post_loss_hook(sub_batch)
losses_td = self._process_loss_hook(sub_batch, losses_td)
losses_detached = self._optimizer_hook(losses_td)  # single backward + step
self._post_optim_hook()
```

This works well for single-phase algorithms (DQN, simple PPO) where all losses are computed once and a single optimizer step follows.

However, TD3 (and other multi-phase algorithms like SAC with auto-tuned entropy, or some model-based methods) requires a fundamentally different optimization order:

1. compute critic loss
2. step critic optimizer
3. compute actor loss using the **updated** critic parameters
4. (delayed) step actor optimizer

As best as I can see in the current design, this either requires overriding `optim_steps()` (inheritance-only customization) or relying on hacky usage of mid-optimization hook stages (`post_loss`, `process_loss`, `optimizer`).

## Solution

Introduce `OptimizationStepper` as a single, composable object owned by the `Trainer`:
- `Trainer.__init__` accepts `optimization_stepper: OptimizationStepper | None`.
- If not provided, a `DefaultOptimizationStepper` is used (which is the current single-phase optimization behaviour)
- The stepper is responsible for **loss computation + stepping** (including multi-optimizer flows).
- Mid-optimization hook destinations (`post_loss`, `process_loss`, `optimizer`) are removed. Their logic becomes part of the `OptimizationStepper`, which allows for much more flexibility overall.

## Alternatives

1. Override `optim_steps()` in subclasses: requires inheritance for every algorithm variant and breaks composition. Also becomes a bit awkward with Hydra if I want to parametrize this multi-phase behaviour.
2. Add phased optimization stages: this seems like a pretty heavy API and overfits to actor/critic structure.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
